src/systems/physics-system.js

```diff
@@ -1,103 +1,311 @@
+              this.bodyAngularVelocities[uuid] = this.objectMatricesFloatArray[
+                index * BUFFER_CONFIG.BODY_DATA_SIZE + 17
+              ];
+            }
+
+            this.collisions[uuid].length = 0;
+
+            for (let j = 18; j < BUFFER_CONFIG.BODY_DATA_SIZE; j++) {
+              const collidingIndex = this.objectMatricesIntArray[index * BUFFER_CONFIG.BODY_DATA_SIZE + j];
+              if (collidingIndex !== -1) {
+                this.collisions[uuid].push(this.indexToUuid[collidingIndex]);
+              }
+            }
+          }
+
+          this.ammoWorker.postMessage(
+            { type: MESSAGE_TYPES.TRANSFER_DATA, objectMatricesFloatArray: this.objectMatricesFloatArray },
+            [this.objectMatricesFloatArray.buffer]
+          );
         }
-      }
-      const time = performance.now();
-      this.world.step(dt / 1000);
-      this.stepDuration = performance.now() - time;
-      for (let i = 0; i < this.bodies.length; i++) {
-        if (this.bodies[i].type === TYPE.DYNAMIC) {
-          this.bodies[i].syncFromPhysics();
+
+        /* DEBUG RENDERING */
+        if (this.debugEnabled) {
+          const index = window.Atomics.load(this.debugIndex, 0);
+          if (index !== 0) {
+            this.debugGeometry.attributes.position.needsUpdate = true;
+            this.debugGeometry.attributes.color.needsUpdate = true;
+          }
+          this.debugGeometry.setDrawRange(0, index);
+          window.Atomics.store(this.debugIndex, 0, 0);
         }
       }
-    }
+    };
+  })();
+
+  addBody(object3D, options) {
+    this.workerHelpers.addBody(this.nextBodyUuid, object3D, options);
+    this.object3Ds[this.nextBodyUuid] = object3D;
+    this.bodyOptions[this.nextBodyUuid] = options;
+    this.collisions[this.nextBodyUuid] = [];
+    this.bodyLinearVelocities[this.nextBodyUuid] = 0;
+    this.bodyAngularVelocities[this.nextBodyUuid] = 0;
+    return this.nextBodyUuid++;
   }
 
-  addBody(body) {
-    this.bodies.push(body);
+  updateBody(uuid, options) {
+    this.bodyOptions[uuid] = options;
+    this.workerHelpers.updateBody(uuid, options);
   }
 
-  removeBody(body) {
-    const idx = this.bodies.indexOf(body);
+  removeBody(uuid) {
+    delete this.indexToUuid[this.uuidToIndex[uuid]];
+    delete this.uuidToIndex[uuid];
+    delete this.object3Ds[uuid];
+    delete this.bodyOptions[uuid];
+    delete this.collisions[uuid];
+    delete this.bodyLinearVelocities[uuid];
+    delete this.bodyAngularVelocities[uuid];
+    const idx = this.bodyUuids.indexOf(uuid);
     if (idx !== -1) {
-      this.bodies.splice(idx, 1);
+      this.bodyUuids.splice(idx, 1);
+    }
+    this.workerHelpers.removeBody(uuid);
+  }
+
+  addShapes(bodyUuid, mesh, options) {
+    if (mesh) {
+      const scale = new THREE.Vector3();
+      scale.setFromMatrixScale(mesh.matrixWorld);
+    }
+    this.workerHelpers.addShapes(bodyUuid, this.nextShapeUuid, mesh, options);
```

Comment:
Any concern over cloning all their vertices before posting them to the worker? If we could avoid this, that seems good. https://github.com/InfiniteLee/three-ammo/pull/2/files#diff-168726dbe96b3ce427e7fedce31bb0bcR43